mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/search/ValueMatcher.java

```diff
@@ -0,0 +1,92 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.vault.search;
+
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.BiFunction;
+
+public interface ValueMatcher<T> {
+
+    boolean matches(T referenceValue);
+
+    class SingleValueMatcher<T> implements ValueMatcher<T> {
+
+        private static final BiFunction<String, String, Boolean> CONTAINS = (testedValue, referenceValue) -> referenceValue.contains(testedValue);
```

Comment:
I really wonder if all of these class could not be some simple lambdas in Criterion factory...